Default dtype flip (#1776)
google/flax: docs/flip/1777-default-dtype.md

# FLIP: Default dtypes

- Start Date: 2022-01-11
- FLIP PR: [#1776](https://github.com/google/flax/pull/1776)
- FLIP Issue: [#1777](https://github.com/google/flax/issues/1777)

## Summary

> This FLIP proposes to replace the default dtype which is currently hard-coded to float32, and instead use the JAX type promotion results to derive a default dtype from the input and parameters of a layer.

Reviewer: This is a bit confusing: it sounds like there is a "hard coded default dtype", but it's rather that we have a "dtype that defaults to float32" and the output is hard-coded to be cast into that dtype.

Also consider using something like `module.dtype` to make the difference with `jnp.dtype`.

Reviewer 2: I agree, "hard-coded" makes it sound like the user can't change it.

Reviewer 3: Before #1739, whose design agrees with what was proposed in this FLIP, you could not change the parameter dtype of  the parameters (if not by specifying a custom initialiser with the dtype you wanted). I believe that's what is meant here by hardcoded?

Author: rephrased

Author: rephrased

## Motivation

> Currently, linen Modules always produce float32 outputs regardless of input and parameter dtypes. Half-precision types like float16 and bfloat16 are supported by explicitly passing the half-precision type to each layer. The way this is currently implemented is that each layer has a dtype argument with float32 as the default value. The layer guarantees that this dtype will be the return type of the result returned by `__call__`.

Reviewer: nit: we always uppercase Linen

> The current behavior is problematic and results in silent bugs especially for dtypes that do not fit inside float32 (complex, float64). Also the linen dtype behavior is significantly different from how NumPy and by extension JAX handle dtypes.

Reviewer: It feels like that's never a good idea. I'm wondering if we should always raise an error if one of the `xs` tree leaves is complex but `dtype` is not. Users should maybe remove imaginary part by themselves if that's really what they want to do.
(Maybe it's a contrived example, but I can imagine cases where layers have their `dtype` set by parent modules based on assumptions without complex numbers in mind)

Reviewer 2: Andreas: FYI I believe this entire FLIP was motivated mainly for complex numbers, and what you are describing is indeed the main issue that users using complex numbers ran into.

Reviewer 3: @andsteing One of the big issues we face when using Linen with NetKet/Quantum Mechanics is that we usually want to specify the dtype of the parameters in the module (real/complex) but we want the output type to be decided by type promotion.
This is because often our input type can change and both be real/complex depending on what quantities we are computing, but the model itself does not change.

To put it differently, 95% of the time, in our use-cases, we treat layers such as dense as if they are 'matrices' and we are doing `Dense @ input`: we do not want to downcast types, and `Dense.dtype` for us matches the dtype of the internal data (aka, the parameters). 
Of course this is not consistent with flax of having modules be callables, where users do `Dense(input)` and where, now, `Dense.dtype` should reflect the output dtype.

This is why @jheek proposed to have a default `Dense.dtype = None`, which will follow default numpy type promotion, and if you want to override that the user has to specify it explicitly.

Reviewer: I agree `Dense.dtype = None` would work as expected.

But I was wondering: Is there *ever* a case where `Dense.dtype = float32` would make sense with a complex input (or parameters)? If not, then we could throw an error in that case (for extra safety).

Reviewer 3: Theoretically I'd like if that could be interpreted as single-precision and mean that complex128 would be casted to complex64.

However this rule is not used anywhere in Numpy (unfortunately) and therefore as a stop gap I think it's better to error.

Regardless, erroring now gives us the possibility of changing this behavior in the future.

Author: @andsteing we could indeed manually throw an error but I left it out of this FLIP (it could easily be added in a separate PR). The reasoning is that we would be more strict than NumPy in this case. And this silent bug is better solved at the JAX level where it's easy to catch and would apply more generally (will add to discussion).

> ### Dtypes in JAX

Author: Added

> ### Dtypes in JAX

Author: done

> JAX uses a NumPy inspired [dtype promotion](https://github.com/google/jax/blob/main/jax/_src/dtypes.py) mechanism as explained [here](https://jax.readthedocs.io/en/latest/type_promotion.html?highlight=lattice#type-promotion-semantics). The type promotion rules are summarized by the following type lattice:

Reviewer 2: ```suggestion
JAX uses a NumPy-inspired [dtype promotion](https://github.com/google/jax/blob/main/jax/_src/dtypes.py) mechanism as explained [here](https://jax.readthedocs.io/en/latest/type_promotion.html?highlight=lattice#type-promotion-semantics). The type promotion rules are summarized by the following type lattice:
```

![JAX type promotion lattice](https://jax.readthedocs.io/en/latest/_images/type_lattice.svg)

## Dtypes in Linen

> Beside input arguments, state and in particular parameters could affect dtype promotion. For example: We might feed a float64 input to a Dense layer with float32 parameters. Currently the result would be truncated to float32. If the input is a complex number the result is even worse because the imaginary part will be silently dropped when casting to float32.

Reviewer 4: ```suggestion
Beside input arguments, state and in particular parameters could affect dtype promotion. For example: we might feed a float64 input to a Dense layer with float32 parameters. Currently the result would be truncated to float32. If the input is a complex number the result is even worse because the imaginary part will be silently dropped when casting to float32.
```

> By using the dtype promotion rules already available in JAX we can avoid this issue. A public api is available called `jax.numpy.result_dtype(*args)` which returns the dtype that JAX would promote the given arguments in accordance with the type promotion lattice. For linen layers the arguments would be the layer inputs together with the parameters so for example this would be inputs, kernel, and bias for a linear layer.

Reviewer 4: ```suggestion
By using the dtype promotion rules already available in JAX we can avoid this issue. A public API is available called `jax.numpy.result_dtype(*args)` which returns the dtype that JAX would promote the given arguments in accordance with the type promotion lattice. For linen layers the arguments would be the layer inputs together with the parameters so for example this would be inputs, kernel, and bias for a linear layer.
```

> By using the dtype promotion rules already available in JAX we can avoid this issue. A public api is available called `jax.numpy.result_dtype(*args)` which returns the dtype that JAX would promote the given arguments in accordance with the type promotion lattice. For linen layers the arguments would be the layer inputs together with the parameters so for example this would be inputs, kernel, and bias for a linear layer.

Reviewer 2: ```suggestion
By using the dtype promotion rules already available in JAX we can avoid this issue. A public api is available called `jax.numpy.result_dtype(*args)`, which returns the dtype that JAX would promote the given arguments to, in accordance with the type promotion lattice. For Linen layers the arguments would be the layer inputs together with the parameters. For example, for a linear layer this would be inputs, kernel, and bias.
```

# Implementation

A simplified example implementation:

```python
def promote_arrays(*xs, dtype):
 if dtype is None:

>    dtype = jnp.result_type(*jax.tree_leaves(xs))

Reviewer: I assume there are never performance issues with this statement. Just wanted to mention it so you can close it and say there are none so it's not overlooked.

Author: No this is cheap

 return jax.tree_map(lambda x: jnp.asarray(x, dtype), xs)

Dtype = Any
class Dense(nn.Module):
 features: int
 kernel_init: Callable
 bias_init: Callable

>  dtype: Optional[Dtype] = None

Reviewer 2: I wonder if the current API isn't a bit confusing now. From a user perspective, do you think they will get what the difference is between `param_dtype` and `dtype`? An alternative is to rename `dtype` to `output_dtype`, or is that inaccurate?

Reviewer: I would vote for keeping it as proposed, because
1. it's the same naming as before
2. it hints at that the computation is done in `dtype` as opposed to transforming the outputs only

Author: That's a big change from a backward's compatibility perspective. I also agree it lacks the computational hint. Anyway, I will add the naming to the discussion section but I will leave it out of the FLIP because we cannot handle too large backward incompatibility in a single change

 param_dtype: Dtype = jnp.float32

 @nn.compact
 def __call__(self, x):
   kernel = self.param("kernel", 
                       self.kernel_init,
                       (x.shape[-1], self.features), self.param_dtype)
   bias = self.param("bias", self.bias_init, (self.features,), self.param_dtype)
   x, kernel, bias = promote_arrays(x, kernel, bias, dtype=self.dtype)
   return x @ kernel + bias
```

> ## Half precision types

Reviewer 2: ```suggestion
## Half-precision dtypes
```

> Some layers don’t work with half-precision types internally. For example: The normalization layers currently compute mean and variance in float32 even when a half precision type is specified to avoid numerical issues. We can replicate this behavior by calling result\_dtype with a dummy argument that has the minimum precision for the sub computation to work correctly.

Reviewer 2: ```suggestion
Some layers don’t work with half-precision dtypes internally. For example: The normalization layers currently compute mean and variance in float32 even when a half-precision dtype is specified to avoid numerical issues. We can replicate this behavior by calling result_dtype with a dummy argument that has the minimum precision for the sub computation to work correctly.
```

> Some layers don’t work with half-precision types internally. For example: The normalization layers currently compute mean and variance in float32 even when a half precision type is specified to avoid numerical issues. We can replicate this behavior by calling result\_dtype with a dummy argument that has the minimum precision for the sub computation to work correctly.

Reviewer 2: I suppose JAX also has a utility function to find the minimal dtype in the lattice? It may be useful mentioning that here. Or can we simply use `jax.results_type`?

## Backward compatibility

> This proposal causes some layers to behave differently in cases where the dtype is not specified to a linen Module. By default, parameters are in float32. Therefore, passing in half or float32 precision inputs will cause a float32 dtype and no functional differences with current behavior.

Reviewer: Could you elaborate a bit more in detail which are the cases where the behavior changes, and which of these cases are problematic.

Maybe even consider a checklist of the problematic cases that we could point users to during the update if they want to make sure they're not negatively affected by the change.

Author: done

> When passing complex or float64 precision, the result will no longer truncate the imaginary component or the precision. The silent truncation is problematic and has caused user complaints. Therefore, this change can be considered a bugfix.

Reviewer 2: It would be useful to point to the user complaint. I think Philip wrote some issue / discussion somewhere right?

Reviewer 3: I summarized my 'complaints' in https://github.com/google/flax/issues/805#issuecomment-981468837 . This can surely be linked.

Author: done

## Corner cases

> How to handle non-parameter output state on a case-by-case basis:

Reviewer: I'm having a hard time parsing "non-parameter output state". Is it outputs, updated state, or both? In either case "non-parameter" seems to be redundant.

Author: clarified

**Autoregressive decoding cache**

> Currently, only attention implements autoregressive cache and the stored key and value mirror the dtype of the key and value passed to the layer. Forcing the cache dtype to be the same as the output dtype could result in reduced precision during cached decoding vs uncached. This seems undesirable. Decision: keep the current behavior.

Reviewer 2: ```suggestion
Currently, only attention implements autoregressive caching and the stored key and value mirror the dtype of the key and value passed to the layer. Forcing the cache dtype to be the same as the output dtype could result in reduced precision during cached decoding vs uncached. This seems undesirable. Decision: keep the current behavior.
```

> Currently, only attention implements autoregressive cache and the stored key and value mirror the dtype of the key and value passed to the layer. Forcing the cache dtype to be the same as the output dtype could result in reduced precision during cached decoding vs uncached. This seems undesirable. Decision: keep the current behavior.

Reviewer 2: So that means that attention will not work correctly for complex numbers, right? I would be useful to check with @PhilipVinc whether they use this, because if so it may be a better decision to add a new attention layer that does work correctly.

Reviewer 3: Nobody has done it yet but we're working on it! @jwnys has been adapting some flax layers to fulfill our needs so he can comment on it.

Reviewer 5: We had issues like this trying to implement attention, indeed (by me, I mean Tom @tvieijra). This was discussed here in issue #1705

Reviewer 6: The problem we encountered was in the implementation of DenseGeneral, that attention layers use for transforming the inputs.  Essentially, the dtype of the parameter initializers was not used to initialize the parameters, resulting in incosistencies between dtypes in our code.  This issue was already fixed in issue #1705.  

We don't use the decoding cache right now, so we don't have problems with the issue mentioned here.  I think the decoding mechanism is still valuable for our purposes, especially as autoregressive sampling is often used.  I think having this work with complex numbers will definitely help people working with attention and autoregressive sampling.

> **Batch statistics**

Reviewer: Above you mention "normalization layers currently compute mean and variance in float32" and here you only talk about batch norm - is that on purpose?

Reviewer 2: I agree this seems to apply to both layer norm and batch norm. Perhaps it is useful to link to this section where you discuss the normalization layers above?

Author: No batchnorm is different because it also stores batch statistics which the other norm layers do not

BatchNorm layers are often used with a half precision output dtype. However, calculating statistics is by default always done in float32 to avoid numerical precision issues and over/underflow for float16. With float64 this would actually cause a downcast so we should now use `np.promote_types(float32, dtype)` such that the precision is at least float32. The running batch statistics will be stored with the same dtype for consistency.

**Complex number support**

Currently, our complex number support is brittle because the default behavior is to truncate the output to the real part. This issue will be fixed by the automatic type promotion proposed in this FLIP. However, some layers require some additional thought to extend to complex number correctly:

> 1. Normalization layers use the complex conjugate to calculate norms instead of normal squaring

Reviewer 2: I don't understand what point 1 actually says. Do you mean this needs more discussion? I think it would be good to explain this in a bit more detail.

```suggestion
1. Normalization layers use the complex conjugate to calculate norms instead of normal squaring.
```

2. Attention: It’s not exactly clear how the dot product and softmax are defined in this case. Raise an error on complex inputs.

> 3. Recurrent layers: might require special gating / activation functions to function correctly but these can be specified by the user

Reviewer 2: ```suggestion
3. Recurrent layers: might require special gating / activation functions to function correctly, but these can be specified by the user.
```

# Discussion

> TODO

Reviewer: Is the idea to summarize our discussions from the FLIP in the end in this section?
(Or were you going to write discussions that we should review as well once it's there but it's not yet written?)

Author: done
